Query frontend now discards cached results that are outside of tenant's configured TTL

pstibrany opened the pull request:
#### What this PR does

This PR adds query timestamp into cached results. This allows query-frontend to discard any cached results that are no longer valid due to tenant's configured cache results TTL.

This fixes a problem described in https://github.com/grafana/mimir/issues/4383#issuecomment-1457771080: changing TTL for cached results used by specific tenant didn't invalidate previously cached results that are now outside of TTL.

All entries cached by previous version of Mimir will be invalidated by this change.

#### Which issue(s) this PR fixes or relates to

Fixes #4385

#### Checklist

- [x] Tests updated
- [na] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on pkg/frontend/querymiddleware/model.proto:
```diff
@@ -87,6 +87,9 @@ message Extent  {
   google.protobuf.Any response = 5 [(gogoproto.jsontag) = "response"];
+  // Unix timestamp (millis) when the query with this response was executed.
+  // If the response is combination of multiple queries over time, queryTime is timestamp of oldest query.
+  int64 queryTime = 6;
```
[nit] In other protos (e.g. `mimir.proto`), we call it "timestamp_ms". You may want to consider `query_timestamp_ms` for consistency.

pracucci commented on pkg/frontend/querymiddleware/split_and_cache.go:
```diff
@@ -365,23 +375,59 @@ func (s *splitAndCacheMiddleware) fetchCacheExtents(ctx context.Context, keys []
+}
+
 // storeCacheExtents stores the extents for given key in the cache.
 func (s *splitAndCacheMiddleware) storeCacheExtents(key string, tenantIDs []string, extents []Extent) {
```
I just realised: we have some logic checking `len(extents) > 0` but what's the purpose of caching no extents? Should we just have a circuit breaker at the beginning of this function to skip the function execution if there are no extents?

pracucci reviewed (approved):
Good job Peter! The design LGTM. I just left few minor comments.

pstibrany commented on pkg/frontend/querymiddleware/split_and_cache.go:
```diff
@@ -365,23 +375,59 @@ func (s *splitAndCacheMiddleware) fetchCacheExtents(ctx context.Context, keys []
+		// Filter out extents that are outside TTL.
+		for ix := range resp.Extents {
+			if resp.Extents[ix].QueryTime == 0 {
+				// If we don't know the query time, it's too old.
+				extentsOutOfTTL++
+				continue
```
I haven't considered it. It would mean that "per-tenant TTL" feature only starts working correctly 7 days after updating your Mimir installation. On the other hand, people can always flush cache manually, if needed, so I'm fine with that approach too.

pstibrany replied:
I've implemented this, but the previous logic was easier to understand, especially when merging multiple extents.

pstibrany commented on pkg/frontend/querymiddleware/split_and_cache.go:
```diff
@@ -365,23 +375,59 @@ func (s *splitAndCacheMiddleware) fetchCacheExtents(ctx context.Context, keys []
+				continue
+			}
+
+			usedTTL := ttl
+			if oooWindow > 0 && extentWithinOOOWindow(&resp.Extents[ix], now, oooWindow) {
+				usedTTL = ttlForExtentsInOOOWindow
+			}
```
We can do that. Then it really becomes "getTTLForExtent", which is fine. When storing extents we will only call it on latest extent, or we can save extents by TTL.

pstibrany commented:
Addressed all feedback in https://github.com/grafana/mimir/pull/4439/commits/e1e88551d65d81122915098c99d301f966e7cb27
- renamed QueryTime to query_timestamp_ms
- keep extents that have query_timestamp_ms == 0, added tests for this case
- skipping storing extents if len(extents) == 0
- added `getTTLForExtent`
